server/apm/apm-server/apm-service/src/main/java/io/holoinsight/server/apm/server/service/impl/ApmMetricMaterializer.java

```diff
@@ -0,0 +1,148 @@
+    List<TenantOps> tenantOpsList = tenantOpsMapper.selectList(wrapper);
+    if (CollectionUtils.isNotEmpty(tenantOpsList)) {
+      Map<String, List<String>> finalIndexGroups = indexGroups;
+      tenantOpsList.forEach(tenantOps -> {
+        String tenant = tenantOps.getTenant();
+        Map<String, MetricDefine> metrics = metricsManager.getMetricDefines();
+        if (metrics != null) {
+          metrics.forEach((metricName, metricDefine) -> {
+            List<String> groups = finalIndexGroups.get(metricDefine.getIndex());
+            EXECUTOR.submit(() -> {
+              try {
+                log.info(
+                    "[apm] ready to materialize metric, tenant={}, metric={}, start={}, end={}, step={}, groups={}",
+                    tenant, metricName, start, end, STEP, groups);
+                MetricValues metricValues = apmMetricStorage.queryMetric(tenant, metricName,
+                    new Duration(start, end, STEP), null, groups);
+                log.info(
+                    "[apm] query metric success, tenant={}, metric={}, start={}, end={}, step={}, groups={}, series={}",
+                    tenant, metricName, start, end, STEP, groups,
+                    metricValues == null ? 0 : CollectionUtils.size(metricValues.getValues()));
+                WriteMetricsParam param = new WriteMetricsParam();
+                param.setTenant(tenant);
+                param.setPoints(new ArrayList<>());
+                AtomicInteger pointCnt = new AtomicInteger(0);
+                if (metricValues != null && CollectionUtils.isNotEmpty(metricValues.getValues())) {
+                  metricValues.getValues().forEach(value -> {
+                    value.getValues().forEach((t, v) -> {
+                      WriteMetricsParam.Point point = new WriteMetricsParam.Point();
+                      point.setMetricName(metricName);
+                      point.setTags(value.getTags());
+                      point.setTimeStamp(t);
+                      point.setValue(v);
+                      param.getPoints().add(point);
+                      pointCnt.incrementAndGet();
+                    });
+                  });
+                }
+                metricStorage.write(param).subscribe(unused -> {
+                  log.info(
+                      "[apm] materialize metric success, tenant={}, metric={}, start={}, end={}, step={}, points={}",
+                      tenant, metricName, start, end, STEP, pointCnt.get());
+                }, error -> {
+                  log.error(
+                      "[apm] materialize metric failed, tenant={}, metric={}, start={}, end={}, step={}, points={}, "
+                          + "msg={}",
+                      tenant, metricName, start, end, STEP, pointCnt.get(), error.getMessage(),
+                      error);
+                });
+              } catch (Exception e) {
+                log.error("[apm] materialize failed, msg={}", e.getMessage(), e);
```

Comment:
No need to add error message on the log message. Because it is added by the last error parameter.

Reply:
Fixed.